lib/testing.cf

```diff
@@ -0,0 +1,359 @@
+
+    inform_mode::
+      "$(this.bundle): report summary: $(out)";
+    tap.inform_mode::
+      "$(this.bundle): TAP report summary: $(tap_tests_info)";
+      "$(this.bundle): TAP report results summary: $(tap_results_info)";
+}
+
+# @brief Bail out in TAP format **immediately**
+# @param reason the bailout reason
+#
+# **See also:** `testing_tap_report`, `testing_ok_if`, `testing_todo`, `testing_skip`, and `testing_ok`
+bundle agent testing_tap_bailout(reason)
+{
+  reports:
+      "$(const.n)Bail out! $(reason)";
+}
+
+bundle agent testing_usage_example
+# @brief Simple demo of testing_junit_report and testing_tap_report testing.cf usage
+#
+# You can run it like this: `cf-agent -K ./testing.cf -b testing_usage_example`
+# Or for extra debugging, you can run it like this: `cf-agent -KI ./testing.cf -b testing_usage_example`
+#
+# You can either use `tap` as the `format` parameter for any testing bundle, in
+# which case you get immediate TAP output, OR you can use anything else, in
+# which case you can still get TAP output but at the end.
+#
+# So your use cases are:
+#
+# * format=jUnit, then testing_junit_report(''): all jUnit to STDOUT, output at end
+# * format=TAP, then testing_tap_report(''): all TAP to STDOUT, immediate output
+# * format=delayed_TAP, then testing_tap_report(MYFILE): all TAP to MYFILE, output at end
+# * format=jUnit, then testing_jUnit_report(MYFILE): all jUnit to MYFILE, output at end
+{
+  classes:
+      "reported_class" scope => "namespace";
+
+  methods:
+      "" usebundle => testing_ok("reported_class", "ok message", "TAP");
+      "" usebundle => testing_ok_if("missing_class", "missing message", "error1", "error trace 1", "TAP");
+      "" usebundle => testing_ok_if("missing_class2", "missing message2", "error2", "error trace 2", "TAP");
+      "" usebundle => testing_skip("skipped_class", "we skipped this", "TAP");
+      "" usebundle => testing_todo("todo_class", "we need to do this", "TAP");
+
+      # output the reports to some files
+      # "" usebundle => testing_junit_report("/var/cfengine/outputs/junit.xml");
+      # "" usebundle => testing_tap_report("/var/cfengine/outputs/tap.txt");
+
+      # output the reports to STDOUT
+      "" usebundle => testing_junit_report('');
+      "" usebundle => testing_tap_report('');
+}
+
+########################################################################
+# Other useful bodies and bundles
+
+body classes testing_diff_classes_generic(x)
+# This is a mix of body classes u_kept_successful_command from update.cf
+# and body classes classes_generic from lib/3.6/common.cf
+# Diffing two files is quite common to test if two different versions
+# of the same policy produced the same result. This version of classes
+# generic will help.
+#
+# From man diff:
+# Exit status is 0 if inputs are the  same,  1  if
+# different, 2 if trouble.
+#
+# @brief Define `x` prefixed/suffixed with promise outcome, Set command to "kept" instead of "repaired" if it returns 0
```

Comment:
Instead of having a full additional body can't we use inherit_from => classes_generic or more preferable results